stdlib/dynarray.mli

```diff
@@ -0,0 +1,615 @@
+    elements to the end of the array.
+
+    We say that an index into a dynarray [a] is valid if it is in
+    [0 .. length a - 1] and invalid otherwise.
+*)
+
+val create : unit -> 'a t
+(** [create ()] is a new, empty array. *)
+
+val make : int -> 'a -> 'a t
+(** [make n x] is a new array of length [n], filled with [x].
+
+    @raise Invalid_argument if [n < 0].
+*)
+
+val init : int -> (int -> 'a) -> 'a t
+(** [init n f] is a new array [a] of length [n],
+    such that [get a i] is [f i]. In other words,
+    the elements of [a] are [f 0], then [f 1],
+    then [f 2]... and [f (n - 1)] last, evaluated
+    in that order.
+
+    This is similar to {!Array.init}.
+
+    @raise Invalid_argument if [n < 0].
+*)
+
+val get : 'a t -> int -> 'a
+(** [get a i] is the [i]-th element of [a], starting with index [0].
+
+    @raise Invalid_argument if the index is invalid *)
+
+val set : 'a t -> int -> 'a -> unit
+(** [set a i x] sets the [i]-th element of [a] to be [x].
+
+    [i] must be a valid index. [set] does not add new elements to the
+    array -- see {!add_last} to add an element.
+
+    @raise Invalid_argument if the index is invalid. *)
+
+val length : 'a t -> int
+(** [length a] is the number of elements in the array. *)
+
+val is_empty : 'a t -> bool
+(** [is_empty a] is [true] if [a] is empty, that is, if [length a = 0]. *)
+
+val copy : 'a t -> 'a t
+(** [copy a] is a shallow copy of [a], a new array
+    containing the same elements as [a]. *)
+
+
+(** {1:adding Adding elements}
+
+    Note: all operations adding elements raise [Invalid_argument] if the
+    length needs to grow beyond {!Sys.max_array_length}. *)
+
+val add_last : 'a t -> 'a -> unit
+(** [add_last a x] adds the element [x] at the end of the array [a]. *)
+
+val append_array : 'a t -> 'a array -> unit
+(** [append_array a b] adds all elements of [b] at the end of [a],
+    in the order they appear in [b].
+
+    For example:
+    {[
+      let a = Dynarray.of_list [1;2]
+      let () = Dynarray.append_array a [|3; 4|]
+      let () = assert (Dynarray.to_list a = [1; 2; 3; 4])
+    ]}
+*)
+
+val append_list : 'a t -> 'a list -> unit
+(** Like {!append_array} but with a list. *)
+
+val append : 'a t -> 'a t -> unit
+(** [append a b] is like [append_array a b],
+    but [b] is itself a dynamic arreay instead of a fixed-size array.
+
+    Beware! Calling [append a a] iterates on [a] and adds elements to
+    it at the same time; it is a programming error and fails with
+    [Invalid_argument].
+*)
+
+val append_seq : 'a t -> 'a Seq.t -> unit
+(** Like {!append_array} but with a sequence.
+
+    Beware! Calling [append_seq a (to_seq a)] is unspecified and may
+    result in an infinite loop, see the {!append} comment above.
+*)
+
+val append_iter :
+  'a t ->
+  (('a -> unit) -> 'iter -> unit) ->
+  'iter -> unit
+(** [append_iter a iter x] adds each element of [x] to the end of [a].
+    This is [iter (add_last a) x].
```

Comment:
Replaces all the menagerie of append_* functions you just saw before which in my opinion should not exist.

Reply:
I decided to keep this function.

Reply:
This is an explicitly third-order function (a first?) whose use is not much shorter than its definition. The argument `(('a -> unit) -> 'iter -> unit)` basically means `{List,Queue,Bytes,Array}.iter` – it doesn't match `Map.S.iter` or `Hashtbl.iter`, or the `iter` functions with named arguments from `base`. The iterator `('a -> unit) -> unit` representation of a container might be a nicer abstraction here, it's at least far more common.

Reply:
> the `iter` functions with named arguments from `base`

That's unfortunate for them but it's a well known fact that pointless labelling doesn't fare well with higher order functions. 

> This is an explicitly third-order function (a first?)

Since 14e8353 there is also `Format.pp_print_iter`.

Regarding @314eter's comment I agree that `'iter` is more confusing than enlighenting here. I'm fine with `'container` but my own eyes would rather be used to simply have a `'b` here. 

However regarding the doc string I would avoid using `iterator [f]` in the doc string. `f` is a function no need to add new terminology here. I would rather tweak the existing doc string to:

```
[append_iter a f v] iterates over the elements of [v] using [f] and
adds them to [a]. This is [f (add_last a) v]. Typically used 
with [M.iter] functions like {!List.iter}.
```

Reply:
I was hesitating between `'b` and `'container`. A `'b` will look more familiar to people that are used to reading OCaml interfaces, and `'container` makes the meaning more clear. But I think your proposal is still clear enough. Especially since it doesn't have to be a "container", it can be any kind of element generator, or even something like `append_iter a iter_dividers_of 24`.